### PR TITLE
[aoti] Add initial custom op support

### PR DESCRIPTION
**Pull request by angelayi**
Re-land of https://github.com/pytorch/pytorch/pull/125242

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -580,6 +580,7 @@ cc_library(
         "@eigen",
         "@fbgemm//:fbgemm_src_headers",
         "@fmt",
+        "@nlohmann",
```
Comment (swolchok) [inline BUILD.bazel:583]: adding a new third-party dep could be a separate PR

File: test/inductor/test_aot_inductor.py [hunk 2]
```diff
@@ -2910,7 +2925,21 @@ def forward(self, x, y):
 common_utils.instantiate_parametrized_tests(AOTInductorTestsTemplate)
 
 
-class AOTInductorTestABICompatibleCpu(TestCase):
+class CustomTestCase(TestCase):
```
Comment (swolchok) [inline test/inductor/test_aot_inductor.py:2928]: nit: all code we write is custom. How about CustomOpTestCase?

File: test/inductor/test_aot_inductor_utils.py [hunk 3]
```diff
@@ -49,6 +49,15 @@ def compile(
...
             )
 
+        if IS_FBCODE:
+            from deeplearning.aot_inductor.extern_node_thrift_serializer import (
+                thrift_serializer,
+            )
+            if options is None:
+                options = {"extern_node_serializer": thrift_serializer}
+            else:
+                options["extern_node_serializer"] = thrift_serializer
```
Comment (swolchok) [inline test/inductor/test_aot_inductor_utils.py:59]: nit: you can avoid repeating yourself like so

```
if options is None:
    options = {}
options["extern_node_serializer"] = thrift_serializer
```

(yes it's slightly slower, but that battle was already lost when we started writing Python)

File: torch/csrc/inductor/aoti_runner/model_container_runner.cpp [hunk 4]
```diff
@@ -46,6 +47,18 @@ AOTIModelContainerRunner::AOTIModelContainerRunner(
   get_call_spec_func_ = reinterpret_cast<decltype(get_call_spec_func_)>(
       model_so_->sym("AOTInductorModelContainerGetCallSpec"));
 
+  // Hack to find the json file name from the model so file
+  size_t lastindex = model_so_path.find_last_of(".");
+  const std::string& json_filename =
```
Comment (swolchok) [inline torch/csrc/inductor/aoti_runner/model_container_runner.cpp:52]: this only happens to work because of an arcane C++ quirk (temporary lifetime extension -- https://abseil.io/tips/107); `std::string json_filename = ...` would be clearer

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+}
+} // namespace
+
+namespace torch {
+namespace aot_inductor {
+
+void OSSProxyExecutor::prefill_stack_with_static_arguments(
+    int index,
+    at::TypePtr schema_arg_type,
+    json serialized_arg,
```
Comment (swolchok) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:21]: I've never used nlohmann before, but surely the `json` type isn't cheap to copy and therefore should be passed by const reference?

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        index++, schema_arg.real_type(), arg, op_kernel);
+  }
+
+  // TODO: prefill default values
+}
+
+// Populates op_kernel.outputs_
+void OSSProxyExecutor::get_output_info_from_serialized(
+    const std::vector<c10::Argument>& schema_returns,
+    json serialized_node,
```
Comment (swolchok) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:64]: ditto pass by reference

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+  // Access data
+  for (auto const& serialized_extern_node : json_obj["nodes"]) {
+    auto const& serialized_node = serialized_extern_node["node"];
+
+    const std::string& target = serialized_node["target"];
+
+    std::string opName;
+    std::string overloadName;
+    size_t pos = target.find(".");
```
Comment (swolchok) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:150]: nit: use a character literal, not a string literal (`'.'`)

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    if (pos == std::string::npos) {
+      opName = target;
+      overloadName = "";
+    } else {
+      // There should be no more periods
+      size_t pos2 = target.find(".", pos);
```
Comment (swolchok) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:156]: ditto character literal

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+      extern_node_index < static_cast<int>(op_kernels_.size()),
+      "Invalid extern node index");
+  OpKernel& op_kernel = op_kernels_[extern_node_index];
+
+  std::vector<c10::IValue> stack = op_kernel.stack_;
+  auto& dynamic_args = op_kernel.dynamic_args_;
+
+  int tensor_id = 0;
+  int int_id = 0;
+  for (size_t i = 0; i < dynamic_args.size(); i++) {
```
Comment (swolchok) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:195]: nit: you don't need the value of `i`, so why not use a range-for loop?

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+struct DynamicArg {
+  DynamicArg(
+      int arg_index,
+      DynamicArgType arg_type,
+      int length,
+      json serialized_arg_val)
+      : arg_index(arg_index),
+        arg_type(arg_type),
+        length(length),
+        serialized_arg_val(serialized_arg_val) {}
```
Comment (swolchok) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:38]: missing std::move

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+  TORCH_CHECK(serialized_arg.size() == 1);
+  std::string serialized_arg_type = serialized_arg.begin().key();
+  auto& serialized_arg_val = serialized_arg.begin().value();
+
+  switch (schema_arg_type->kind()) {
+    case c10::TypeKind::TensorType: {
+      TORCH_CHECK(serialized_arg_type == "as_tensor");
+      stack.emplace_back();
+      dynamic_args.emplace_back(
+          index, DynamicArgType::TensorType, 1, serialized_arg_val);
```
Comment (swolchok) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:35]: missing std::move(serialized_arg_val)

more importantly, IIUC we are going to parse JSON once at startup and then use it in a potentially-performance-critical path, potentially very many times. Shouldn't we extract the data into plain old structs to save memory and hash-table-lookups later?

**Review by swolchok — COMMENTED** [review]
not well-versed in inductor; just some general comments + higher-level issue of JSON parsing

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp (hunk 11 above)
Comment (angelayi) [reply torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp]: Ah yes that makes sense! I can address this in a follow up!

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+#include <ATen/core/ivalue.h>
+#include <c10/macros/Export.h>
+#include <nlohmann/json.hpp>
+#include <torch/csrc/inductor/aoti_torch/c/shim.h>
+#include <torch/csrc/inductor/aoti_torch/proxy_executor.h>
+
+using json = nlohmann::json;
+
+namespace torch {
+namespace aot_inductor {
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:13]: Nit
```suggestion
namespace torch::aot_inductor {
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+enum DynamicArgType : int {
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:15]: Not to leak it in a global namespace
```suggestion
enum class DynamicArgType : int {
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+      const std::vector<c10::Argument>& schema_returns,
+      const json& serialized_node,
+      OpKernel& op_kernel);
+
+  std::vector<OpKernel> op_kernels_;
+  std::unique_ptr<c10::Device> device_;
+};
+
+} // namespace aot_inductor
+} // namespace torch
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:99]:
```suggestion
} // namespace torch::aot_inductor
```

File: torch/csrc/inductor/aoti_runner/model_container_runner.cpp [hunk 15]
```diff
@@ -2,8 +2,17 @@
 #include <ATen/DynamicLibrary.h>
 
 #include <torch/csrc/inductor/aoti_runner/model_container_runner.h>
+#include <torch/csrc/inductor/aoti_torch/oss_proxy_executor.h>
 #include <torch/csrc/inductor/aoti_torch/tensor_converter.h>
 
+#if __has_include("filesystem")
```
Comment (malfet) [inline torch/csrc/inductor/aoti_runner/model_container_runner.cpp:8]:
```suggestion
// TODO: Investigate why this is necessary, but fixes build problems in FRL
#if __has_include("filesystem")
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+using json = nlohmann::json;
+
+namespace {
+at::Tensor* tensor_handle_to_tensor_pointer(AtenTensorHandle handle) {
+  return reinterpret_cast<at::Tensor*>(handle);
+}
+} // namespace
+
+namespace torch {
+namespace aot_inductor {
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:16]:
```suggestion
namespace torch::aot_inductor {
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+  TORCH_CHECK(
+      tensor_id == num_tensors,
+      "Mismatch between tensors consumed and num_tensors, got tensor_id = ",
+      tensor_id,
+      ", expected num = ",
+      num_tensors);
+}
+
+} // namespace aot_inductor
+} // namespace torch
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:266]:
```suggestion
} // namespace torch::aot_inductor
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 18]
```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <ATen/core/dispatch/Dispatcher.h>
+#include <ATen/core/ivalue.h>
+#include <c10/macros/Export.h>
+#include <nlohmann/json.hpp>
+#include <torch/csrc/inductor/aoti_torch/c/shim.h>
+#include <torch/csrc/inductor/aoti_torch/proxy_executor.h>
+
+using json = nlohmann::json;
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:10]: Not using in headers please
```suggestion
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+      arg_type == DynamicArgType::ListTensorType ||
+      arg_type == DynamicArgType::ListOptionalTensorType;
+}
+
+struct DynamicArg {
+  DynamicArg(
+      int arg_index,
+      DynamicArgType arg_type,
+      int length,
+      json serialized_arg_val)
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:34]:
```suggestion
      nlohmann::json serialized_arg_val)
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+  for (const auto& schema_return : schema_returns) {
+    if (schema_return.type()->kind() == c10::TypeKind::TensorType) {
+      at::Tensor* tensor =
+          tensor_handle_to_tensor_pointer(flatten_tensor_args[tensor_id++]);
+      *tensor = stack[index++].toTensor();
+      // TODO: handle tensor list returns
+    } else {
+      TORCH_CHECK(
+          false,
+          "NYI: Unsupported return type for schema: ",
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:252]: What NYI stands for?

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+      : arg_index(arg_index),
+        arg_type(arg_type),
+        length(length),
+        serialized_arg_val(std::move(serialized_arg_val)) {}
+  int arg_index;
+  DynamicArgType arg_type;
+  int length;
+  json serialized_arg_val;
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:42]:
```suggestion
  nlohmann::json serialized_arg_val;
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+  OpKernel(const std::string& target, const c10::OperatorHandle& op_handle)
+      : target_(target), op_handle_(op_handle) {}
+
+  std::string target_;
+  c10::OperatorHandle op_handle_;
+  std::vector<DynamicArg> dynamic_args_;
+  std::vector<DynamicArg> outputs_;
+  std::vector<c10::IValue> stack_;
+
+  int num_output_tensors() const {
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:55]: Why signed value?

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+ private:
+  void prefill_stack_with_static_arguments(
+      int index,
+      at::TypePtr schema_arg_type,
+      const json& thrift_arg,
+      OpKernel& op_kernel);
+
+  void get_input_info_from_serialized(
+      const std::vector<c10::Argument>& schema_args,
+      const json& serialized_node,
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:86]:
```suggestion
      const nlohmann::json& serialized_node,
```

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+      OpKernel& op_kernel);
+
+  void get_output_info_from_serialized(
+      const std::vector<c10::Argument>& schema_returns,
+      const json& serialized_node,
```
Comment (malfet) [inline torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:91]:
```suggestion
      const nlohmann::json& serialized_node,
```

**Review by malfet — APPROVED** [review]

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp (hunk 20 above)
Comment (angelayi) [reply torch/csrc/inductor/aoti_torch/oss_proxy_executor.cpp:252]: not yet implemented!

File: torch/csrc/inductor/aoti_torch/oss_proxy_executor.h (hunk 22 above)
Comment (angelayi) [reply torch/csrc/inductor/aoti_torch/oss_proxy_executor.h:55]: I just followed [these signed values](https://github.com/pytorch/pytorch/blob/f093cd40864d5c8d4ceecb72fc9828a9fcb94a4e/torch/csrc/inductor/aoti_torch/proxy_executor.h#L15-L20) 😅 not sure if those should also be changed to size_t
